frame/staking/src/lib.rs

```diff
@@ -853,8 +965,55 @@ decl_module! {
 
 		fn deposit_event() = default;
 
-		fn on_initialize() {
+		/// Does the following:
+		///
+		/// 1. potential storage migration
+		/// 2. sets `ElectionStatus` to `Triggered(now)` where `now` is the block number at which
+		///    the election window has opened. The offchain worker, if applicable, will execute at
+		///    the end of the current block. `submit_election_solution` will accept solutions from
+		///    this block until the end of the era.
+		fn on_initialize(now: T::BlockNumber) {
 			Self::ensure_storage_upgraded();
+			if
+				// if we don't have any ongoing offchain compute.
+				Self::era_election_status() == ElectionStatus::None &&
+				// and an era is about to be changed.
+				Self::is_current_session_final()
+			{
+				let next_session_change =
+					T::NextSessionChange::estimate_next_session_change(now);
+				if let Some(remaining) = next_session_change.checked_sub(&now) {
+					if remaining <= T::ElectionLookahead::get() && !remaining.is_zero() {
+						// Set the flag to make sure we don't waste any compute here in the same era
+						// after we have triggered the offline compute.
+						<EraElectionStatus<T>>::put(
+							ElectionStatus::<T::BlockNumber>::Open(now)
+						);
+						debug::native::info!(
+							target: "staking",
+							"detected a good block ({}) to trigger offchain election. Submission will \
+							be allowed from the next block.",
+							now,
+						);
+					}
+				}
+			}
+		}
+
+		/// Check if the current block number is the one at which the election window has been set
+		/// to open. If so, it runs the offchain worker code.
+		fn offchain_worker(now: T::BlockNumber) {
+			debug::RuntimeLogger::init();
+			// runs only once.
+			if Self::era_election_status() == ElectionStatus::<T::BlockNumber>::Open(now) {
```

Comment:
That's not really a reliable way of running offchain workers. We don't have guarantees that they will run on every block, you should rather check that current block is greater and additionally have a local storage (offchain db) entry that prevents running twice. This seems to become a pattern, might be good to encapsulate that in some util (CC @Xanewok )

Reply:
Hmm what do you mean running twice? when I limit my mind to when juts a node is running and is in sync with the tip, this should be fine. 

But if you mean while we re-sync etc. Yeah, actually that should have been prevented. I assume we do it in im-online, will look there.